crates/qasm3/src/lib.rs

```diff
@@ -150,11 +154,147 @@ pub fn load(
+        }
+        if let Ok(val) = kw.get_item("disable_constants") {
+            options.disable_constants = val.extract::<bool>()?;
+        }
+        if let Ok(val) = kw.get_item("allow_aliasing") {
+            options.allow_aliasing = val.extract::<bool>()?;
+        }
+        if let Ok(val) = kw.get_item("indent") {
+            options.indent = val.extract::<String>()?;
+        }
+    }
+    let circuit_data = circuit
+        .getattr("_data")?
+        .downcast::<CircuitData>()?
+        .borrow();
+
+    let islayout = !circuit.getattr("layout")?.is_none();
+
+    let exporter = exporter::Exporter::new(
+        options.includes,
+        options.basis_gates,
+        options.disable_constants,
+        options.allow_aliasing,
+        options.indent,
+    );
+
+    let stream = exporter.dumps(&circuit_data, islayout).map_err(|err| {
+        QASM3ImporterError::new_err(format!(
+            "failed to export circuit using qasm3.dumps_experimental: {:?}",
+            err
+        ))
+    })?;
+
+    Ok(stream)
+}
+
+#[pyfunction]
+#[pyo3(signature = (circuit,stream, /, kwargs=None))]
+pub fn dump(
+    _py: Python,
+    circuit: &Bound<PyAny>,
+    stream: &Bound<PyAny>,
+    kwargs: Option<&Bound<PyDict>>,
+) -> PyResult<()> {
+    let mut options = DumpOptions::default();
+
+    if let Some(kw) = kwargs {
+        let kw = kw.as_ref();
+        if let Ok(val) = kw.get_item("includes") {
+            options.includes = val.extract::<Vec<String>>()?;
+        }
+        if let Ok(val) = kw.get_item("basis_gates") {
+            options.basis_gates = val.extract::<Vec<String>>()?;
+        }
+        if let Ok(val) = kw.get_item("disable_constants") {
+            options.disable_constants = val.extract::<bool>()?;
+        }
+        if let Ok(val) = kw.get_item("allow_aliasing") {
+            options.allow_aliasing = val.extract::<bool>()?;
+        }
+        if let Ok(val) = kw.get_item("indent") {
+            options.indent = val.extract::<String>()?;
+        }
+    }
+    let circuit_data = circuit
+        .getattr("_data")?
+        .downcast::<CircuitData>()?
+        .borrow();
+
+    let islayout = !circuit.getattr("layout")?.is_none();
+
+    let exporter = exporter::Exporter::new(
+        options.includes,
+        options.basis_gates,
+        options.disable_constants,
+        options.allow_aliasing,
+        options.indent,
+    );
+
+    let mut output = Vec::new();
+    exporter
+        .dump(&circuit_data, islayout, &mut output)
+        .map_err(|err| {
+            QASM3ImporterError::new_err(format!(
+                "failed to export circuit using qasm3.dump_experimental: {:?}",
+                err
+            ))
+        })?;
+
+    let output_str = String::from_utf8(output)
+        .map_err(|e| QASM3ImporterError::new_err(format!("invalid utf-8 output: {:?}", e)))?;
+
+    stream.call_method1("write", (output_str,))?;
+
+    Ok(())
+}
```

Comment:
Accepting a text-mode stream in the Python-space `dump` isn't going to transition well to Rust - if we're given an open Python-space stream, we almost certainly should handle it by creating a full string from Rust and then copying it across to Python space, unless this is a performance pipeline we _really_ want to optimise, in which case we should make a Rust object that implements `Write`, and buffers writes out to the underlying Python stream so we aren't paying a huge runtime penalty for going Rust -> Python all the time (like how general file-backed streams also buffer writes).

For the time being, I feel like the whole Rust-space `dump` is unnecessary, and we could achieve the same thing just in the Python-space wrapper, with it calling Rust-space `dumps`.